File: 1-arraysum/arraysum.py

arr = [3, 5, -4, 8, 11, 1, -1, -16]
sum = 10


def one():
    # simple 2 pointers
    # time n^2
    for i in range(len(arr)):
        for j in range(i+1, len(arr)):
            if arr[i] + arr[j] == sum:
                print("{} {}".format(arr[i], arr[j]))


def two():

    # dict
    # time n space n
    dic = {}

    for i in arr:

        if dic.get(sum - i) != None:
            print("{} {}".format(sum - i, i))
        else:
            dic[i] = 1


def three():

    # sorting, 2 pointers
    # time nlogn 
    sArr = sorted(arr)

    i,j =0, len(arr)-1

    while( i < j ):
        curr = sArr[i] + sArr[j]

        if(curr == sum):
            print("{} {}".format(sArr[i], sArr[j]))
            break
        elif(curr < sum):
            i+=1
        else:
            j-=1



print("Started..!!")

one()
two()
three()

File: 1-arraysum/arraysum.java
import java.util.Arrays;
import java.util.Hashtable;
import java.util.Map;

import apple.laf.JRSUIUtils.Tree;

class ArraySum {

public static void main(String[] args){
    
    System.out.println("I started..!!");
    int[] arr = {3,5,-4,8,11,1,-1,-16};
    int sum = 10; //19
    one(arr,sum);
    second(arr, sum);
    third(arr, sum);
}

    private static void third(int[] arr, int sum){

        //using sorting and 2 pointers
        //time nlogn, space c

        Arrays.sort(arr);

        int i =0;
        int j = arr.length-1;

        while( i<j ){

            int curr = arr[i] + arr[j];

            if(curr == sum){
                System.out.println(arr[i] + " " + arr[j]);
                break;
            }
            else if(curr < sum){
                i++;
            }
            else{
                j--;
            }
        }

    }

    private static void second(int[] arr, int sum){

        //using hashmap
        //time n, space n

        Map<Integer,Boolean> map = new Hashtable<>();

        for(int i:arr){

            if (map.containsKey(sum - i)){
                System.out.println(i + " , "  + (sum-i));
            }
            else{
                map.put(i, true);
            }

        }

    }

    private static void one(int[] arr, int sum) {

        int n = arr.length;

        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {

                if (arr[i] + arr[j] == sum) {
                    System.out.println(arr[i] + " " + arr[j]);
                }
            }
        }

    }

}